{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "37932dc9-9d41-480d-b8a3-618890410df3",
   "metadata": {},
   "outputs": [],
   "source": [
    "#This block to load all libraries needed\n",
    "library(\"DEXSeq\")\n",
    "library(\"tidyr\")\n",
    "library(\"qvalue\")\n",
    "library(\"edgeR\")\n",
    "library(\"dplyr\")\n",
    "library(\"reshape2\")\n",
    "library(\"EnhancedVolcano\")\n",
    "library(\"sva\")\n",
    "library(\"Rtsne\")\n",
    "library(\"ggplot2\")\n",
    "library(\"pheatmap\")\n",
    "library(\"org.Hs.eg.db\")\n",
    "library(\"readr\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "40096413-a34f-46f7-996b-f29fb701515b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   count                      Subgroup\n",
      "1    771                     Antisense\n",
      "2   1225               Long intergenic\n",
      "3     39                      Intronic\n",
      "4    202          Long non-coding RNAs\n",
      "5     12            Small nuclear RNAs\n",
      "6      0                transfer RNAs \n",
      "7    157                     Divergent\n",
      "8    452                      MicroRNA\n",
      "9      0               Mitochondrially\n",
      "10     1 Nuclear-encoded mitochondrial\n",
      "11     7       Overlapping transcripts\n",
      "12     0              Piwi-interacting\n",
      "13     0                     Ribosomal\n",
      "14     2             Ro60-associated Y\n",
      "15     0     Small Cajal body-specific\n",
      "16     0                    Small NF90\n",
      "17    82           Small nucleolar RNA\n",
      "18     8                    Variant U1\n",
      "19     0                         Vault\n",
      "20 24016                         24016\n",
      "   count                      Subgroup\n",
      "1    584                     Antisense\n",
      "2    954               Long intergenic\n",
      "3     20                      Intronic\n",
      "4    166          Long non-coding RNAs\n",
      "5      5            Small nuclear RNAs\n",
      "6      0                transfer RNAs \n",
      "7    120                     Divergent\n",
      "8    289                      MicroRNA\n",
      "9      0               Mitochondrially\n",
      "10     0 Nuclear-encoded mitochondrial\n",
      "11     6       Overlapping transcripts\n",
      "12     0              Piwi-interacting\n",
      "13     0                     Ribosomal\n",
      "14     1             Ro60-associated Y\n",
      "15     0     Small Cajal body-specific\n",
      "16     0                    Small NF90\n",
      "17    65           Small nucleolar RNA\n",
      "18     0                    Variant U1\n",
      "19     0                         Vault\n",
      "20 19349                         19349\n",
      "   count                      Subgroup\n",
      "1    591                     Antisense\n",
      "2    991               Long intergenic\n",
      "3     25                      Intronic\n",
      "4    169          Long non-coding RNAs\n",
      "5      2            Small nuclear RNAs\n",
      "6      0                transfer RNAs \n",
      "7    133                     Divergent\n",
      "8    266                      MicroRNA\n",
      "9      0               Mitochondrially\n",
      "10     0 Nuclear-encoded mitochondrial\n",
      "11     7       Overlapping transcripts\n",
      "12     0              Piwi-interacting\n",
      "13     0                     Ribosomal\n",
      "14     0             Ro60-associated Y\n",
      "15     0     Small Cajal body-specific\n",
      "16     0                    Small NF90\n",
      "17    57           Small nucleolar RNA\n",
      "18     0                    Variant U1\n",
      "19     0                         Vault\n",
      "20 19663                         19663\n",
      "   count                      Subgroup\n",
      "1    549                     Antisense\n",
      "2    963               Long intergenic\n",
      "3     32                      Intronic\n",
      "4    158          Long non-coding RNAs\n",
      "5      4            Small nuclear RNAs\n",
      "6      0                transfer RNAs \n",
      "7    121                     Divergent\n",
      "8    249                      MicroRNA\n",
      "9      0               Mitochondrially\n",
      "10     0 Nuclear-encoded mitochondrial\n",
      "11     5       Overlapping transcripts\n",
      "12     0              Piwi-interacting\n",
      "13     0                     Ribosomal\n",
      "14     1             Ro60-associated Y\n",
      "15     0     Small Cajal body-specific\n",
      "16     0                    Small NF90\n",
      "17    51           Small nucleolar RNA\n",
      "18     2                    Variant U1\n",
      "19     0                         Vault\n",
      "20 18810                         18810\n"
     ]
    }
   ],
   "source": [
    "#This block will grab each method's count files, then process all the exons to generate the venn diagram.\n",
    "\n",
    "#grab method's files\n",
    "\n",
    "#ExCy\n",
    "Excy.countfiles = list.files('./Excy/PC2/', pattern=\"txt$\", full.names=TRUE)\n",
    "\n",
    "#ExoEasy\n",
    "EE.countfiles = list.files('./ExoEasy/PC2/', pattern=\"txt$\", full.names=TRUE)\n",
    "\n",
    "#Fujifilm\n",
    "FF.countfiles = list.files('./Fujifilm/PC2/', pattern=\"txt$\", full.names=TRUE)\n",
    "\n",
    "#UC\n",
    "UC.countfiles = list.files('./UC/PC2/', pattern=\"txt$\", full.names=TRUE)\n",
    "\n",
    "##NEXT is the GFF\n",
    "flattenedFile = list.files('.', pattern=\"gff$\", full.names=TRUE)\n",
    "\n",
    "###NEXT IS THE METADATA\n",
    "metadata = read.csv('ExCy_metadata_dexseq.csv')\n",
    "rownames(metadata) <- metadata$SampleID\n",
    "metadata <- metadata[,-1]\n",
    "\n",
    "print('Files are found, success')\n",
    "\n",
    "####annotation files\n",
    "exomega <- as.data.frame(read.csv('ExoMEGA.txt'))\n",
    "ncrna_data <- read_tsv('non-coding-rna.txt', show_col_types=FALSE)\n",
    "ncrna <-  ncrna_data %>% select(c(2,13))\n",
    "mRNA <- read_tsv('VESICLEPEDIA_PROTEIN_MRNAS_5.1.txt')\n",
    "\n",
    "###########Next blocks are to tabulate the count files using feature counts\n",
    "\n",
    "##the steps in BLOCK order (increases with a # per step) for each sample block:\n",
    "\n",
    "# grab the sample row\n",
    "# read the sample row, then detect all exons in the first column for the column. \n",
    "# set the column vector to the row names\n",
    "# drop the unneeded column vector\n",
    "# tag all exons corresponding to the same transcript \n",
    "# filter all transcript counts less than 10 and clean up intermediate ID column\n",
    "# Combine all exons corresponding to the same transcript with melt, group_by, summarize, then shift from long form to wide form data structure with dcast.\n",
    "\n",
    "\n",
    "##############################################\n",
    "\n",
    "\n",
    "#ExCy\n",
    "Excy.metadata <- metadata[1,] \n",
    "Excy.dxd = read.table(Excy.countfiles, header=FALSE) %>% dplyr::filter(grepl(\"ENSG\", V1, fixed=TRUE))\n",
    "Excy.dxd <- Excy.dxd %>% mutate(Gene = substr(V1, 1,15))\n",
    "Excy.cf <- Excy.dxd %>%  dplyr::filter(V2 > 10) %>% select(-V1)\n",
    "Excy.cfs <-  melt(Excy.cf, id.vars = \"Gene\") %>% group_by(Gene, variable) %>% summarise(sum_value=sum(value)) %>% dcast(Gene ~ variable, value.var=\"sum_value\")\n",
    "rownames(Excy.cfs) <- Excy.cfs$Gene \n",
    "\n",
    "\n",
    "#map gene names to HUGO\n",
    "annot.df <- mapIds(org.Hs.eg.db, keys = row.names(Excy.cfs), column = \"SYMBOL\", keytype = \"ENSEMBL\") %>% as.data.frame() %>% rename(SYMBOL=1)\n",
    "annot.df$SYMBOL <- toupper(annot.df$SYMBOL)\n",
    "Excy.cfs$gene_id <- annot.df$SYMBOL \n",
    "\n",
    "#filter out non-mapped genes and combine duplicate\n",
    "Excy.gcfs <- Excy.cfs %>% dplyr::filter(!grepl(\"NA\", gene_id, fixed= TRUE)) %>% select(-Gene)\n",
    "Excy.gpcfs <-  melt(Excy.gcfs, id.vars = \"gene_id\") %>% group_by(gene_id, variable) %>% summarise(sum_value=sum(value)) %>% dcast(gene_id ~ variable, value.var=\"sum_value\")\n",
    "Excy.gpcfs <- Excy.gpcfs[complete.cases(Excy.gpcfs), ]\n",
    "write.csv(Excy.gpcfs, 'Excy_before_map_venn_PC2.csv')\n",
    "\n",
    "##Vesiclepedia mapping\n",
    "Excy.exomega <- as.data.frame(Excy.gpcfs[Excy.gpcfs$gene_id %in% exomega[,1] ,])\n",
    "rownames(Excy.exomega) <- Excy.exomega$gene_id\n",
    "write.csv(Excy.exomega, 'Excy_after_map_PC2.csv')\n",
    "\n",
    "#now map to mRNA\n",
    "mRNA_Excy <- Excy.exomega[rownames(Excy.exomega) %in% mRNA$mRNA, ]\n",
    "write.csv(mRNA_Excy, 'Excy_mRNA_PC2.csv')\n",
    "Excy.em <- mRNA_Excy %>% select(-gene_id)\n",
    "\n",
    "\n",
    "#######################################################################################\n",
    "\n",
    "#EE\n",
    "EE.metadata <- metadata[2,]\n",
    "EE.dxd = read.table(EE.countfiles, header=FALSE) %>% dplyr::filter(grepl(\"ENSG\", V1, fixed=TRUE))\n",
    "EE.dxd <- EE.dxd %>% mutate(Gene = substr(V1, 1,15))\n",
    "EE.cf <- EE.dxd %>%  dplyr::filter(V2 > 10) %>% select(-V1)\n",
    "EE.cfs <-  melt(EE.cf, id.vars = \"Gene\") %>% group_by(Gene, variable) %>% summarise(sum_value=sum(value)) %>% dcast(Gene ~ variable, value.var=\"sum_value\")\n",
    "rownames(EE.cfs) <- EE.cfs$Gene\n",
    "\n",
    "#map gene names to HUGO\n",
    "annot.df <- mapIds(org.Hs.eg.db, keys = row.names(EE.cfs), column = \"SYMBOL\", keytype = \"ENSEMBL\") %>% as.data.frame() %>% rename(SYMBOL=1)\n",
    "annot.df$SYMBOL <- toupper(annot.df$SYMBOL)\n",
    "EE.cfs$gene_id <- annot.df$SYMBOL \n",
    "\n",
    "#filter out non-mapped genes and combine duplicate\n",
    "EE.gcfs <- EE.cfs %>% dplyr::filter(!grepl(\"NA\", gene_id, fixed= TRUE)) %>% select(-Gene)\n",
    "EE.gpcfs <-  melt(EE.gcfs, id.vars = \"gene_id\") %>% group_by(gene_id, variable) %>% summarise(sum_value=sum(value)) %>% dcast(gene_id ~ variable, value.var=\"sum_value\")\n",
    "EE.gpcfs <- EE.gpcfs[complete.cases(EE.gpcfs), ]\n",
    "write.csv(EE.gpcfs, 'EE_before_map_PC2.csv')\n",
    "\n",
    "##Vesiclepedia mapping\n",
    "EE.exomega <- as.data.frame(EE.gpcfs[EE.gpcfs$gene_id %in% exomega[,1] ,])\n",
    "write.csv(EE.exomega, 'EE_after_map_PC2.csv')\n",
    "rownames(EE.exomega) <- EE.exomega$gene_id\n",
    "mRNA_EE <- EE.exomega[rownames(EE.exomega) %in% mRNA$mRNA ,]\n",
    "write.csv(mRNA_EE, 'EE_mRNA_PC2.csv')\n",
    "EE.em <- mRNA_EE %>% select(-gene_id)\n",
    "\n",
    "#######################################################################################\n",
    "\n",
    "#FF\n",
    "FF.metadata <- metadata[3,]\n",
    "FF.dxd = read.table(FF.countfiles, header=FALSE) %>% dplyr::filter(grepl(\"ENSG\", V1, fixed=TRUE))\n",
    "FF.dxd <- FF.dxd %>% mutate(Gene = substr(V1, 1,15))\n",
    "FF.cf <- FF.dxd %>% dplyr::filter(V2 > 10) %>% select(-V1)\n",
    "FF.cfs <-  melt(FF.cf, id.vars = \"Gene\") %>% group_by(Gene, variable) %>% summarise(sum_value=sum(value)) %>% dcast(Gene ~ variable, value.var=\"sum_value\")\n",
    "rownames(FF.cfs) <- FF.cfs$Gene\n",
    "\n",
    "#map gene names to HUGO\n",
    "annot.df <- mapIds(org.Hs.eg.db, keys = row.names(FF.cfs), column = \"SYMBOL\", keytype = \"ENSEMBL\") %>% as.data.frame() %>% rename(SYMBOL=1)\n",
    "annot.df$SYMBOL <- toupper(annot.df$SYMBOL)\n",
    "FF.cfs$gene_id <- annot.df$SYMBOL \n",
    "\n",
    "#filter out non-mapped genes and combine duplicate\n",
    "FF.gcfs <- FF.cfs %>% dplyr::filter(!grepl(\"NA\", gene_id, fixed= TRUE)) %>% select(-Gene)\n",
    "FF.gpcfs <-  melt(FF.gcfs, id.vars = \"gene_id\") %>% group_by(gene_id, variable) %>% summarise(sum_value=sum(value)) %>% dcast(gene_id ~ variable, value.var=\"sum_value\")\n",
    "FF.gpcfs <- FF.gpcfs[complete.cases(FF.gpcfs), ]\n",
    "write.csv(FF.gpcfs, 'FF_venn_before_map_PC2.csv')\n",
    "\n",
    "##Vesiclepedia mapping\n",
    "FF.exomega <- as.data.frame(FF.gpcfs[FF.gpcfs$gene_id %in% exomega[,1] ,])\n",
    "rownames(FF.exomega) <- FF.exomega$gene_id\n",
    "write.csv(FF.exomega, 'FF_after_PC2_map.csv')\n",
    "mRNA_FF <- FF.exomega[rownames(FF.exomega) %in% mRNA$mRNA, ]\n",
    "write.csv(mRNA_FF, 'FF_mRNA_PC2.csv')\n",
    "FF.em <- mRNA_FF %>% select(-gene_id)\n",
    "\n",
    "#######################################################################################\n",
    "\n",
    "#UC\n",
    "UC.metadata <- metadata[4,]\n",
    "UC.dxd = read.table(UC.countfiles, header=FALSE) %>% dplyr::filter(grepl(\"ENSG\", V1, fixed=TRUE))\n",
    "UC.dxd <- UC.dxd %>% mutate(Gene = substr(V1, 1,15))\n",
    "UC.cf <- UC.dxd %>% dplyr::filter(V2 > 10) %>% select(-V1)\n",
    "UC.cfs <-  melt(UC.cf, id.vars = \"Gene\") %>% group_by(Gene, variable) %>% summarise(sum_value=sum(value)) %>% dcast(Gene ~ variable, value.var=\"sum_value\")\n",
    "rownames(UC.cfs) <- UC.cfs$Gene\n",
    "\n",
    "#\n",
    "annot.df <- mapIds(org.Hs.eg.db, keys = row.names(UC.cfs), column = \"SYMBOL\", keytype = \"ENSEMBL\") %>% as.data.frame() %>% rename(SYMBOL=1)\n",
    "annot.df$SYMBOL <- toupper(annot.df$SYMBOL)\n",
    "UC.cfs$gene_id <- annot.df$SYMBOL \n",
    "\n",
    "#filter out non-mapped genes and combine duplicate\n",
    "UC.gcfs <- UC.cfs %>% dplyr::filter(!grepl(\"NA\", gene_id, fixed= TRUE)) %>% select(-Gene)\n",
    "UC.gpcfs <-  melt(UC.gcfs, id.vars = \"gene_id\") %>% group_by(gene_id, variable) %>% summarise(sum_value=sum(value)) %>% dcast(gene_id ~ variable, value.var=\"sum_value\")\n",
    "UC.gpcfs <- UC.gpcfs[complete.cases(UC.gpcfs), ]\n",
    "write.csv(UC.gpcfs, 'UC_venn_before_map_PC2.csv')\n",
    "\n",
    "##Vesiclepedia mapping\n",
    "UC.exomega <- as.data.frame(UC.gpcfs[UC.gpcfs$gene_id %in% exomega[,1] ,])\n",
    "rownames(UC.exomega) <- UC.exomega$gene_id\n",
    "write.csv(UC.exomega, 'UC_after_map_PC2.csv')\n",
    "mRNA_UC <- UC.exomega[rownames(UC.exomega) %in% mRNA$mRNA, ]\n",
    "write.csv(mRNA_UC, 'UC_venn_PC2.csv')\n",
    "UC.em <- mRNA_UC %>% select(-gene_id)\n",
    "\n",
    "print('Samples exons are found and have been successfuly processed for venn diagram analysis \\n please see the csv to do the venn diagram analysis')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "054ead6f-b9b1-4d4f-8385-7fcf9fac770b",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "   count                      Subgroup\n",
      "1    224                     Antisense\n",
      "2    146               Long intergenic\n",
      "3     32                      Intronic\n",
      "4     78          Long non-coding RNAs\n",
      "5      5            Small nuclear RNAs\n",
      "6      0                transfer RNAs \n",
      "7      0                     Divergent\n",
      "8    362                      MicroRNA\n",
      "9      0               Mitochondrially\n",
      "10     0 Nuclear-encoded mitochondrial\n",
      "11     1       Overlapping transcripts\n",
      "12     0              Piwi-interacting\n",
      "13     0                     Ribosomal\n",
      "14     2             Ro60-associated Y\n",
      "15     0     Small Cajal body-specific\n",
      "16     0                    Small NF90\n",
      "17    60           Small nucleolar RNA\n",
      "18     2                    Variant U1\n",
      "19     0                         Vault\n",
      "20 17888                         17888\n",
      "   count                      Subgroup\n",
      "1    152                     Antisense\n",
      "2    104               Long intergenic\n",
      "3     14                      Intronic\n",
      "4     61          Long non-coding RNAs\n",
      "5      5            Small nuclear RNAs\n",
      "6      0                transfer RNAs \n",
      "7      0                     Divergent\n",
      "8    220                      MicroRNA\n",
      "9      0               Mitochondrially\n",
      "10     0 Nuclear-encoded mitochondrial\n",
      "11     1       Overlapping transcripts\n",
      "12     0              Piwi-interacting\n",
      "13     0                     Ribosomal\n",
      "14     1             Ro60-associated Y\n",
      "15     0     Small Cajal body-specific\n",
      "16     0                    Small NF90\n",
      "17    48           Small nucleolar RNA\n",
      "18     0                    Variant U1\n",
      "19     0                         Vault\n",
      "20 15003                         15003\n",
      "   count                      Subgroup\n",
      "1    158                     Antisense\n",
      "2    115               Long intergenic\n",
      "3     20                      Intronic\n",
      "4     63          Long non-coding RNAs\n",
      "5      2            Small nuclear RNAs\n",
      "6      0                transfer RNAs \n",
      "7      0                     Divergent\n",
      "8    201                      MicroRNA\n",
      "9      0               Mitochondrially\n",
      "10     0 Nuclear-encoded mitochondrial\n",
      "11     1       Overlapping transcripts\n",
      "12     0              Piwi-interacting\n",
      "13     0                     Ribosomal\n",
      "14     0             Ro60-associated Y\n",
      "15     0     Small Cajal body-specific\n",
      "16     0                    Small NF90\n",
      "17    48           Small nucleolar RNA\n",
      "18     0                    Variant U1\n",
      "19     0                         Vault\n",
      "20 15186                         15186\n",
      "   count                      Subgroup\n",
      "1    148                     Antisense\n",
      "2    106               Long intergenic\n",
      "3     27                      Intronic\n",
      "4     56          Long non-coding RNAs\n",
      "5      1            Small nuclear RNAs\n",
      "6      0                transfer RNAs \n",
      "7      0                     Divergent\n",
      "8    185                      MicroRNA\n",
      "9      0               Mitochondrially\n",
      "10     0 Nuclear-encoded mitochondrial\n",
      "11     1       Overlapping transcripts\n",
      "12     0              Piwi-interacting\n",
      "13     0                     Ribosomal\n",
      "14     1             Ro60-associated Y\n",
      "15     0     Small Cajal body-specific\n",
      "16     0                    Small NF90\n",
      "17    40           Small nucleolar RNA\n",
      "18     0                    Variant U1\n",
      "19     0                         Vault\n",
      "20 14510                         14510\n"
     ]
    }
   ],
   "source": [
    "##THis block of code is to generate the RNA biotype distributions using the GENCODE database (as a txt file import)####\n",
    "\n",
    "#Load GENCODE by file\n",
    "ncrna_data <- read_tsv('non-coding-rna.txt', show_col_types=FALSE)\n",
    "ncrna <-  ncrna_data %>% select(c(2,13))\n",
    "\n",
    "\n",
    "\n",
    "#Block write the columns to do the dictionary filter\n",
    "\n",
    "##Make the column\n",
    "ExCy.exomega_filt <- cbind(Excy.gpcfs, Genes=Excy.gpcfs$'gene_id')\n",
    "FF.exomega_filt <- cbind(FF.gpcfs, Genes=FF.gpcfs$'gene_id')\n",
    "UC.exomega_filt <- cbind(UC.gpcfs, Genes=UC.gpcfs$'gene_id')\n",
    "EE.exomega_filt <- cbind(EE.gpcfs, Genes=EE.gpcfs$'gene_id')\n",
    "\n",
    "\n",
    "##Compute teh lengths of each list to generate the total RNA count\n",
    "EE_len <- length(EE.exomega_filt$Genes)\n",
    "FF_len <- length(FF.exomega_filt$Genes)\n",
    "UC_len <- length(UC.exomega_filt$Genes)\n",
    "ExCy_len <- length(ExCy.exomega_filt$Genes)\n",
    "\n",
    "\n",
    "#Do the filtering against GENCODE\n",
    "EE_filt <- ncrna[ncrna$'Approved symbol' %in% EE.exomega_filt$Genes ,]\n",
    "FF_filt <- ncrna[ncrna$'Approved symbol' %in% FF.exomega_filt$Genes ,]\n",
    "UC_filt <- ncrna[ncrna$'Approved symbol' %in% UC.exomega_filt$Genes ,]\n",
    "ExCy_filt <- ncrna[ncrna$'Approved symbol' %in% ExCy.exomega_filt$Genes ,]\n",
    "\n",
    "\n",
    "# Initialize the summary dataframe\n",
    "summary_df <- data.frame(count = numeric(0))\n",
    "\n",
    "# Define dataframes and lengths\n",
    "dataframes <- list(ExCy_filt, EE_filt, FF_filt, UC_filt)\n",
    "lengths <- list(ExCy_len, EE_len, FF_len, UC_len)\n",
    "methods <- list('Excy','ExoEasy','Fujifilm','UC')\n",
    "\n",
    "#Biotypes to look for:\n",
    "keywords <- c(\"Antisense\", \"Long intergenic\", \"Intronic\", \"Long non-coding RNAs\", \"Small nuclear RNAs\", \n",
    "              'transfer RNAs ', 'Divergent', 'MicroRNA', 'Mitochondrially', 'Nuclear-encoded mitochondrial',\n",
    "              'Overlapping transcripts', 'Piwi-interacting', 'Ribosomal','Ro60-associated Y', 'Small Cajal body-specific',\n",
    "              'Small NF90', 'Small nucleolar RNA', 'Variant U1', 'Vault'\n",
    "              )\n",
    "\n",
    "# Loop through dataframes\n",
    "for (i in seq_along(dataframes)) {\n",
    "  current_df <- dataframes[[i]]\n",
    "  current_len <- lengths[[i]]\n",
    "  current_method <- methods[[i]]\n",
    "    \n",
    "  # Loop through keywords and calculate counts and percentages\n",
    "  for (keyword in keywords) {\n",
    "    df_counts <- current_df %>%\n",
    "      filter(grepl(keyword, `Group name`, ignore.case = TRUE)) %>%\n",
    "      summarise(count = n()) %>%\n",
    "      mutate(Subgroup = keyword)\n",
    "    \n",
    "    # Append to summary_df\n",
    "    summary_df <- summary_df %>%\n",
    "      bind_rows(df_counts) \n",
    "  }\n",
    "  \n",
    "  # Append length to summary_df\n",
    "  summary_df <- rbind(summary_df, 'Length' = current_len) \n",
    "  \n",
    "  # Print the method and corresponding result\n",
    "  print(current_method)  \n",
    "  print(summary_df)\n",
    " \n",
    "    \n",
    "  # Reinitialize summary_df for the next iteration\n",
    "  summary_df <- data.frame(count = numeric(0))\n",
    "}\n",
    "\n",
    "###########TOTAL RNA DISTRIBUTION AFTER MAPPING\n",
    "ExCy.exomega_filt <- cbind(Excy.exomega, Genes=rownames(Excy.exomega))\n",
    "FF.exomega_filt <- cbind(FF.exomega, Genes=rownames(FF.exomega))\n",
    "UC.exomega_filt <- cbind(UC.exomega, Genes=rownames(UC.exomega))\n",
    "EE.exomega_filt <- cbind(EE.exomega, Genes=rownames(EE.exomega))\n",
    "\n",
    "\n",
    "\n",
    "EE_len <- length(EE.exomega_filt$Genes)\n",
    "FF_len <- length(FF.exomega_filt$Genes)\n",
    "UC_len <- length(UC.exomega_filt$Genes)\n",
    "ExCy_len <- length(ExCy.exomega_filt$Genes)\n",
    "#length(rownames(EE.exomega))\n",
    "\n",
    "EE_filt <- ncrna[ncrna$'Approved symbol' %in% EE.exomega_filt$Genes ,]\n",
    "FF_filt <- ncrna[ncrna$'Approved symbol' %in% FF.exomega_filt$Genes ,]\n",
    "UC_filt <- ncrna[ncrna$'Approved symbol' %in% UC.exomega_filt$Genes ,]\n",
    "ExCy_filt <- ncrna[ncrna$'Approved symbol' %in% ExCy.exomega_filt$Genes ,]\n",
    "\n",
    "# Initialize the summary dataframe\n",
    "summary_df <- data.frame(count = numeric(0))\n",
    "\n",
    "# Define dataframes and lengths\n",
    "dataframes <- list(ExCy_filt, EE_filt, FF_filt, UC_filt)\n",
    "lengths <- list(ExCy_len, EE_len, FF_len, UC_len)\n",
    "methods <- list('Excy','ExoEasy','Fujifilm','UC')\n",
    "\n",
    "#Biotypes to look for:\n",
    "keywords <- c(\"Antisense\", \"Long intergenic\", \"Intronic\", \"Long non-coding RNAs\", \"Small nuclear RNAs\", \n",
    "              'transfer RNAs ', 'Divergent', 'MicroRNA', 'Mitochondrially', 'Nuclear-encoded mitochondrial',\n",
    "              'Overlapping transcripts', 'Piwi-interacting', 'Ribosomal','Ro60-associated Y', 'Small Cajal body-specific',\n",
    "              'Small NF90', 'Small nucleolar RNA', 'Variant U1', 'Vault'\n",
    "              )\n",
    "\n",
    "# Loop through dataframes\n",
    "for (i in seq_along(dataframes)) {\n",
    "  current_df <- dataframes[[i]]\n",
    "  current_len <- lengths[[i]]\n",
    "  current_method <- methods[[i]]\n",
    "    \n",
    "  # Loop through keywords and calculate counts and percentages\n",
    "  for (keyword in keywords) {\n",
    "    df_counts <- current_df %>%\n",
    "      filter(grepl(keyword, `Group name`, ignore.case = TRUE)) %>%\n",
    "      summarise(count = n()) %>%\n",
    "      mutate(Subgroup = keyword)\n",
    "    \n",
    "    # Append to summary_df\n",
    "    summary_df <- summary_df %>%\n",
    "      bind_rows(df_counts) \n",
    "  }\n",
    "  \n",
    "  # Append length to summary_df\n",
    "  summary_df <- rbind(summary_df, 'Length' = current_len) \n",
    "  \n",
    "  # Print the method and corresponding result\n",
    "  print(current_method)  \n",
    "  print(summary_df)\n",
    " \n",
    "    \n",
    "  # Reinitialize summary_df for the next iteration\n",
    "  summary_df <- data.frame(count = numeric(0))\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "53acd7f7-7d34-467a-be96-50fe80ff2d7f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "converting counts to integer mode\n",
      "\n",
      "Warning message in DESeqDataSet(rse, design, ignoreRank = TRUE):\n",
      "\"some variables in design formula are characters, converting to factors\"\n",
      "Warning message:\n",
      "\"attributes are not identical across measure variables; they will be dropped\"\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'Gene'. You can override using the\n",
      "`.groups` argument.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m9989\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m1\u001b[39m\n",
      "\u001b[36m--\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m--------------------------------------------------------\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (1): mRNA\n",
      "\n",
      "\u001b[36mi\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mi\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene_id'. You can override using the\n",
      "`.groups` argument.\n"
     ]
    }
   ],
   "source": [
    "########THis block of code is to generate the heatmap and correlations shown in the main figure and supplement\n",
    "\n",
    "#Same as before, except when dealing with multiple files, now we can use  feature counts and DEXSeq to grab all the files \n",
    "# read the sample rows, then detect all exons in the first column for the column. \n",
    "# set the column vector to the row names\n",
    "# drop the unneeded column vector\n",
    "# tag all exons corresponding to the same transcript \n",
    "# filter all transcript counts less than 10 and clean up intermediate ID column\n",
    "# Combine all exons corresponding to the same transcript with melt, group_by, summarize, then shift from long form to wide form data structure with dcast.\n",
    "\n",
    "#grab file\n",
    "All.countfiles=list.files('./ALL_PC2/', pattern=\"txt$\", full.names=TRUE)\n",
    "\n",
    "\n",
    "##Generate count file\n",
    "All.dxd = DEXSeqDataSetFromHTSeq(countfiles = All.countfiles, sampleData = metadata[1:4,], design = ~ sample + exon, flattenedfile = flattenedFile)\n",
    "All.count <- as.data.frame(featureCounts(All.dxd))\n",
    "\n",
    "#Do the light filtering and aggregation\n",
    "All.count$ID <- rownames(All.count)\n",
    "All.count <- All.count %>% mutate(Gene = substr(ID, 1,15))\n",
    "All.cf <- All.count %>% mutate(row_sum = rowSums(.[,-5:-6])) %>% dplyr::filter(row_sum > 10) %>% select(-ID)\n",
    "All.cfs <-  melt(All.cf, id.vars = \"Gene\") %>% group_by(Gene, variable) %>% summarise(sum_value=sum(value)) %>% dcast(Gene ~ variable, value.var=\"sum_value\")\n",
    "rownames(All.cfs) <- All.cfs$Gene\n",
    "\n",
    "#prepare for annotation\n",
    "exomega <- as.data.frame(read.csv('ExoMEGA.txt'))\n",
    "mRNA <- read_tsv('VESICLEPEDIA_PROTEIN_MRNAS_5.1.txt')\n",
    "\n",
    "##map gene names to HUGO\n",
    "annot.df <- mapIds(org.Hs.eg.db, keys = row.names(All.cfs), column = \"SYMBOL\", keytype = \"ENSEMBL\") %>% as.data.frame() %>% rename(SYMBOL=1)\n",
    "annot.df$SYMBOL <- toupper(annot.df$SYMBOL)\n",
    "All.cfs$gene_id <- annot.df$SYMBOL \n",
    "\n",
    "#filter out non-mapped genes and combine duplicate\n",
    "All.gcfs <- All.cfs %>% dplyr::filter(!grepl(\"NA\", gene_id, fixed= TRUE)) %>% select(-Gene)\n",
    "All.gpcfs <-  melt(All.gcfs, id.vars = \"gene_id\") %>% group_by(gene_id, variable) %>% summarise(sum_value=sum(value)) %>% dcast(gene_id ~ variable, value.var=\"sum_value\")\n",
    "All.gpcfs <- All.gpcfs[complete.cases(All.gpcfs), ]\n",
    "\n",
    "All.exomega <- as.data.frame(All.gpcfs[All.gpcfs$gene_id %in% exomega[,1] ,])\n",
    "rownames(All.exomega) <- All.exomega$gene_id\n",
    "mRNA_All <- All.exomega[rownames(All.exomega) %in% mRNA$mRNA ,]\n",
    "\n",
    "All.exomega <- mRNA_All[,2:5]\n",
    "\n",
    "###\n",
    "print(\"All files have been collated together, successfully processed, and ready for analysis\")\n",
    "print(\"All files have been collated together, successfully processed, and ready for analysis\")\n",
    "###\n",
    "\n",
    "\n",
    "#######ANALYSIS####\n",
    "\n",
    "##Sample correlation\n",
    "\n",
    "#Change names\n",
    "colnames(All.exomega) <- c(\"ExCy\",\"ExoEasy\",\"Fujifilm\",\"UC\")\n",
    "All.corr <- cor(All.exomega)\n",
    "All.crdy <- round(log2(All.corr+1), 2)\n",
    "\n",
    "\n",
    "pheatmap(All.crdy, \n",
    "         fontsize_row=24, \n",
    "         fontsize_col=24, \n",
    "         display_numbers=T, \n",
    "         fontsize_number=24, \n",
    "         border_color=\"white\", \n",
    "         number_color=\"black\", \n",
    "         legend_labels=c(\"0.75\",\"0.80\",\"0.85\",\"0.90\",\"0.95\",\"1.00\"), \n",
    "         legend_breaks=c(0.75,0.8,0.85,0.90,0.95,1), \n",
    "         color=colorRampPalette(c(\"pink\",\"red\"))(6), \n",
    "         cluster_cols = FALSE, \n",
    "         cluster_rows=FALSE, \n",
    "         filename=\"corr_PC2.png\")\n",
    "\n",
    "\n",
    "#COSMIC HEATMAP \n",
    "COSMIC <- as.data.frame(read.csv('COSMIC.txt'))\n",
    "colnames(All.exomega) <- c(\"ExCy\",\"ExoEasy\",\"Fujifilm\",\"UC\")\n",
    "\n",
    "\n",
    "#Key in cosmic pancreas genes to Matrix\n",
    "All.cosmic <- as.data.frame(All.exomega[rownames(All.exomega) %in% COSMIC[,1] ,])\n",
    "\n",
    "pheatmap(log2(All.cosmic+1), \n",
    "         scale='row', \n",
    "         border_color=\"white\", \n",
    "         fontsize=30, \n",
    "         number_color=\"black\", \n",
    "         display_numbers=F, \n",
    "         fontface=\"bold\", \n",
    "         filename=\"cosmic_heatmap_PC2.png\", \n",
    "         angle_col=0, \n",
    "         width=11, \n",
    "         height=9)\n",
    "\n",
    "####THIS BLOCK IS FOR THE BEEPLOT\n",
    "beeplot <- log2(All.exomega+1) %>% cbind(., Genes=rownames(.)) %>% pivot_longer(-Genes, names_to = \"Method\", values_to = \"Expression\")\n",
    "\n",
    "#color vector\n",
    "colours <- c(\"ExCy\" =\"red3\", \"ExoEasy\"=\"orange\", \"Fujifilm\"=\"grey\", \"UC\"=\"blue2\")\n",
    "\n",
    "\n",
    "#plotting\n",
    "options(repr.plot.width = 9, repr.plot.height =5)\n",
    "\n",
    "\n",
    "ggplot(beeplot, aes(x = Method, y = Expression, color = Method)) +\n",
    "  geom_quasirandom(cex=0.5, width=0.5) +\n",
    "  scale_color_manual(values=colours)+\n",
    "  labs(x = NULL, y = \"Log2 mRNA Transcripts\") +\n",
    "  theme_classic() +\n",
    "    theme(axis.line = element_line(color='black'),\n",
    "    plot.background = element_blank(),\n",
    "    panel.grid.major = element_blank(),\n",
    "    panel.grid.minor = element_blank(),\n",
    "    axis.text.x = element_text(size=25, face=\"bold\", color=\"black\"),\n",
    "    axis.text.y = element_text(size=25, color=\"black\"),\n",
    "    text = element_text(size=25, face='bold')\n",
    "          )"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "UFRC R-4.1",
   "language": "R",
   "name": "ir-ufrc-4.1"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
