{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "7d6d9607-9fae-4b84-8352-7538c62e2969",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Loading required package: BiocParallel\n",
      "\n",
      "Loading required package: Biobase\n",
      "\n",
      "Loading required package: BiocGenerics\n",
      "\n",
      "\n",
      "Attaching package: 'BiocGenerics'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    Filter, Find, Map, Position, Reduce, anyDuplicated, append,\n",
      "    as.data.frame, basename, cbind, colnames, dirname, do.call,\n",
      "    duplicated, eval, evalq, get, grep, grepl, intersect, is.unsorted,\n",
      "    lapply, mapply, match, mget, order, paste, pmax, pmax.int, pmin,\n",
      "    pmin.int, rank, rbind, rownames, sapply, setdiff, sort, table,\n",
      "    tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "Loading required package: SummarizedExperiment\n",
      "\n",
      "Loading required package: MatrixGenerics\n",
      "\n",
      "Loading required package: matrixStats\n",
      "\n",
      "\n",
      "Attaching package: 'matrixStats'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:Biobase':\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'MatrixGenerics'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:matrixStats':\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "The following object is masked from 'package:Biobase':\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "Loading required package: GenomicRanges\n",
      "\n",
      "Loading required package: stats4\n",
      "\n",
      "Loading required package: S4Vectors\n",
      "\n",
      "\n",
      "Attaching package: 'S4Vectors'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    I, expand.grid, unname\n",
      "\n",
      "\n",
      "Loading required package: IRanges\n",
      "\n",
      "Loading required package: GenomeInfoDb\n",
      "\n",
      "Loading required package: DESeq2\n",
      "\n",
      "Loading required package: AnnotationDbi\n",
      "\n",
      "Loading required package: RColorBrewer\n",
      "\n",
      "\n",
      "Attaching package: 'tidyr'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:S4Vectors':\n",
      "\n",
      "    expand\n",
      "\n",
      "\n",
      "Loading required package: limma\n",
      "\n",
      "\n",
      "Attaching package: 'limma'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:DEXSeq':\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "The following object is masked from 'package:DESeq2':\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "The following object is masked from 'package:BiocGenerics':\n",
      "\n",
      "    plotMA\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'dplyr'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:AnnotationDbi':\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:GenomicRanges':\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from 'package:GenomeInfoDb':\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:IRanges':\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:S4Vectors':\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following object is masked from 'package:matrixStats':\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following object is masked from 'package:Biobase':\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:BiocGenerics':\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:stats':\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:base':\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'reshape2'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:tidyr':\n",
      "\n",
      "    smiths\n",
      "\n",
      "\n",
      "Loading required package: ggplot2\n",
      "\n",
      "Loading required package: ggrepel\n",
      "\n",
      "Loading required package: mgcv\n",
      "\n",
      "Warning message:\n",
      "\"package 'mgcv' was built under R version 4.3.1\"\n",
      "Loading required package: nlme\n",
      "\n",
      "\n",
      "Attaching package: 'nlme'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:dplyr':\n",
      "\n",
      "    collapse\n",
      "\n",
      "\n",
      "The following object is masked from 'package:IRanges':\n",
      "\n",
      "    collapse\n",
      "\n",
      "\n",
      "This is mgcv 1.9-0. For overview type 'help(\"mgcv-package\")'.\n",
      "\n",
      "Loading required package: genefilter\n",
      "\n",
      "\n",
      "Attaching package: 'genefilter'\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:MatrixGenerics':\n",
      "\n",
      "    rowSds, rowVars\n",
      "\n",
      "\n",
      "The following objects are masked from 'package:matrixStats':\n",
      "\n",
      "    rowSds, rowVars\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Attaching package: 'readr'\n",
      "\n",
      "\n",
      "The following object is masked from 'package:genefilter':\n",
      "\n",
      "    spec\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#This block to load all libraries needed\n",
    "library(\"DEXSeq\")\n",
    "library(\"tidyr\")\n",
    "library(\"qvalue\")\n",
    "library(\"edgeR\")\n",
    "library(\"dplyr\")\n",
    "library(\"reshape2\")\n",
    "library(\"EnhancedVolcano\")\n",
    "library(\"sva\")\n",
    "library(\"Rtsne\")\n",
    "library(\"ggplot2\")\n",
    "library(\"pheatmap\")\n",
    "library(\"org.Hs.eg.db\")\n",
    "library(\"readr\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "c2830f2c-55d3-424b-9249-7098a27e48eb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "converting counts to integer mode\n",
      "\n",
      "Warning message in DESeqDataSet(rse, design, ignoreRank = TRUE):\n",
      "\"some variables in design formula are characters, converting to factors\"\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "Warning message:\n",
      "\"attributes are not identical across measure variables; they will be dropped\"\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'Gene'. You can override using the\n",
      "`.groups` argument.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m9989\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m1\u001b[39m\n",
      "\u001b[36m--\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m--------------------------------------------------------\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (1): mRNA\n",
      "\n",
      "\u001b[36mi\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mi\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene_id'. You can override using the\n",
      "`.groups` argument.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 x 5</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>logFC</th><th scope=col>logCPM</th><th scope=col>LR</th><th scope=col>PValue</th><th scope=col>FDR</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>GAPDHS</th><td>5.910818</td><td>4.135902</td><td>22.58558</td><td>2.009874e-06</td><td>0.01592021</td></tr>\n",
       "\t<tr><th scope=row>RPS12</th><td>7.528071</td><td>3.919328</td><td>15.59221</td><td>7.857784e-05</td><td>0.13846198</td></tr>\n",
       "\t<tr><th scope=row>TFF1</th><td>7.653591</td><td>3.662601</td><td>15.36555</td><td>8.858898e-05</td><td>0.13846198</td></tr>\n",
       "\t<tr><th scope=row>CD247</th><td>1.495815</td><td>7.634337</td><td>15.31186</td><td>9.114229e-05</td><td>0.13846198</td></tr>\n",
       "\t<tr><th scope=row>TMEM69</th><td>7.229293</td><td>3.142136</td><td>15.24700</td><td>9.432628e-05</td><td>0.13846198</td></tr>\n",
       "\t<tr><th scope=row>CDC25C</th><td>2.190536</td><td>6.332131</td><td>15.04672</td><td>1.048822e-04</td><td>0.13846198</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 x 5\n",
       "\\begin{tabular}{r|lllll}\n",
       "  & logFC & logCPM & LR & PValue & FDR\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\tGAPDHS & 5.910818 & 4.135902 & 22.58558 & 2.009874e-06 & 0.01592021\\\\\n",
       "\tRPS12 & 7.528071 & 3.919328 & 15.59221 & 7.857784e-05 & 0.13846198\\\\\n",
       "\tTFF1 & 7.653591 & 3.662601 & 15.36555 & 8.858898e-05 & 0.13846198\\\\\n",
       "\tCD247 & 1.495815 & 7.634337 & 15.31186 & 9.114229e-05 & 0.13846198\\\\\n",
       "\tTMEM69 & 7.229293 & 3.142136 & 15.24700 & 9.432628e-05 & 0.13846198\\\\\n",
       "\tCDC25C & 2.190536 & 6.332131 & 15.04672 & 1.048822e-04 & 0.13846198\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 x 5\n",
       "\n",
       "| <!--/--> | logFC &lt;dbl&gt; | logCPM &lt;dbl&gt; | LR &lt;dbl&gt; | PValue &lt;dbl&gt; | FDR &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| GAPDHS | 5.910818 | 4.135902 | 22.58558 | 2.009874e-06 | 0.01592021 |\n",
       "| RPS12 | 7.528071 | 3.919328 | 15.59221 | 7.857784e-05 | 0.13846198 |\n",
       "| TFF1 | 7.653591 | 3.662601 | 15.36555 | 8.858898e-05 | 0.13846198 |\n",
       "| CD247 | 1.495815 | 7.634337 | 15.31186 | 9.114229e-05 | 0.13846198 |\n",
       "| TMEM69 | 7.229293 | 3.142136 | 15.24700 | 9.432628e-05 | 0.13846198 |\n",
       "| CDC25C | 2.190536 | 6.332131 | 15.04672 | 1.048822e-04 | 0.13846198 |\n",
       "\n"
      ],
      "text/plain": [
       "       logFC    logCPM   LR       PValue       FDR       \n",
       "GAPDHS 5.910818 4.135902 22.58558 2.009874e-06 0.01592021\n",
       "RPS12  7.528071 3.919328 15.59221 7.857784e-05 0.13846198\n",
       "TFF1   7.653591 3.662601 15.36555 8.858898e-05 0.13846198\n",
       "CD247  1.495815 7.634337 15.31186 9.114229e-05 0.13846198\n",
       "TMEM69 7.229293 3.142136 15.24700 9.432628e-05 0.13846198\n",
       "CDC25C 2.190536 6.332131 15.04672 1.048822e-04 0.13846198"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#GFF\n",
    "flattenedFile = list.files('.', pattern=\"gff$\", full.names=TRUE)\n",
    "\n",
    "#METADATA\n",
    "metadata = read.csv('ExCy_metadata_dexseq.csv')\n",
    "rownames(metadata) <- metadata$SampleID\n",
    "metadata <- metadata[,-1]\n",
    "\n",
    "#Entire dataset generation here to do downstream analyses\n",
    "All.countfiles=list.files('./Tumors/', pattern=\"txt$\", full.names=TRUE)\n",
    "\n",
    "#Generate count file\n",
    "All.dxd = DEXSeqDataSetFromHTSeq(countfiles = All.countfiles, sampleData = metadata[c(1:4,13:20,25:28),], design = ~ sample + exon, flattenedfile = flattenedFile )\n",
    "All.count <- as.data.frame(featureCounts(All.dxd))\n",
    "\n",
    "#Do the light filtering and aggregation\n",
    "All.count$ID <- rownames(All.count)\n",
    "All.count <- All.count %>% mutate(Gene = substr(ID, 1,15))\n",
    "All.cf <- All.count %>% mutate(row_sum = rowSums(.[,-17:-18])) %>% dplyr::filter(row_sum > 10) %>% select(-ID)\n",
    "All.cfs <-  melt(All.cf, id.vars = \"Gene\") %>% group_by(Gene, variable) %>% summarise(sum_value=sum(value)) %>% dcast(Gene ~ variable, value.var=\"sum_value\")\n",
    "rownames(All.cfs) <- All.cfs$Gene\n",
    "\n",
    "#prepare for annotation\n",
    "exomega <- as.data.frame(read.csv('ExoMEGA.txt'))\n",
    "mRNA <- read_tsv('VESICLEPEDIA_PROTEIN_MRNAS_5.1.txt')\n",
    "\n",
    "##map gene names to HUGO\n",
    "annot.df <- mapIds(org.Hs.eg.db, keys = row.names(All.cfs), column = \"SYMBOL\", keytype = \"ENSEMBL\") %>% as.data.frame() %>% rename(SYMBOL=1)\n",
    "annot.df$SYMBOL <- toupper(annot.df$SYMBOL)\n",
    "All.cfs$gene_id <- annot.df$SYMBOL \n",
    "\n",
    "#filter out non-mapped genes and combine duplicate\n",
    "All.gcfs <- All.cfs %>% dplyr::filter(!grepl(\"NA\", gene_id, fixed= TRUE)) %>% select(-Gene)\n",
    "All.gpcfs <-  melt(All.gcfs, id.vars = \"gene_id\") %>% group_by(gene_id, variable) %>% summarise(sum_value=sum(value)) %>% dcast(gene_id ~ variable, value.var=\"sum_value\")\n",
    "All.gpcfs <- All.gpcfs[complete.cases(All.gpcfs), ]\n",
    "\n",
    "\n",
    "All.exomega <- as.data.frame(All.gpcfs[All.gpcfs$gene_id %in% exomega[,1] ,])\n",
    "rownames(All.exomega) <- All.exomega$gene_id\n",
    "mRNA_All <- All.exomega[rownames(All.exomega) %in% mRNA$mRNA ,]\n",
    "All.exomega <- mRNA_All[,2:17]\n",
    "\n",
    "#write.table(All.exomega, \"raw_counts_tumors.txt\", sep=\"\\t\")\n",
    "\n",
    "\n",
    "######################################\n",
    "\n",
    "#edgeR\n",
    "vol.metadata <- metadata[c(1:4,13:20,25:28),]\n",
    "method <- relevel(factor(vol.metadata$Method), ref=\"UC\")\n",
    "#condition <- factor(vol.metadata$Condition)\n",
    "#histology <- factor(vol.metadata$Histology)\n",
    "#sample <- factor(vol.metadata$ID)\n",
    "#pathology <- factor(vol.metadata$Pathology)\n",
    "design <- model.matrix(~method)\n",
    "d <- DGEList(counts = All.exomega)\n",
    "d <- calcNormFactors(d)\n",
    "d <- estimateDisp(d, design, robust=TRUE)\n",
    "d_fit <- glmFit(d, design)\n",
    "qlf <- glmLRT(d_fit, coef=2:4)\n",
    "qlf_table <- topTags(qlf, n=Inf)$table\n",
    "\n",
    "#write.csv(qlf_table, 'altogether_tumor_deg_mrna.csv')\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "####################################\n",
    "\n",
    "\n",
    "#prepare for heatmap\n",
    "All.heatmap <- All.exomega\n",
    "Gene_List = rownames(All.exomega)\n",
    "All.heatmap <- cbind(All.heatmap, Gene_List)\n",
    "rownames(All.heatmap) <- rownames(All.exomega)\n",
    "DEG <- topTags(qlf, n=Inf)$table %>% slice(1:14)\n",
    "\n",
    "All_prep_heatmap <- as.data.frame(All.heatmap[All.heatmap$Gene_List %in% rownames(DEG), ])\n",
    "All_prep_heatmap <- All_prep_heatmap[,-17] \n",
    "colnames(All_prep_heatmap) <- vol.metadata$Method\n",
    "labels <- colnames(All_prep_heatmap)\n",
    "\n",
    "colnames(All_prep_heatmap) <- c(1:16)\n",
    "meta <- data.frame(Patient=list(vol.metadata$ID))\n",
    "colnames(meta) <- c(\"PC\")\n",
    "\n",
    "ann_colors=list(PC=c(PC1=\"#edf8fb\", PC2=\"#b2e2e2\", PC3=\"#66c2a4\", PC4=\"#2ca25f\"))\n",
    "\n",
    "#Graphing\n",
    "\n",
    "pheatmap(\n",
    "    log2(All_prep_heatmap+1),\n",
    "    scale='row',\n",
    "    border_color=\"white\",\n",
    "    clustering_method=\"ward.D2\",\n",
    "    fontsize=16,\n",
    "    number_color=\"black\",\n",
    "    display_numbers=F,\n",
    "    show_rownames=T,\n",
    "    cutree_col=4,\n",
    "    fontface=\"bold\",\n",
    "    angle_col=90,\n",
    "    annotation_col=meta,\n",
    "    annotation_colors=ann_colors,\n",
    "    labels_col=labels,\n",
    "    width=12,\n",
    "    height=6,\n",
    "    filename=\"All_tumor_heatmap_update.png\",\n",
    "    breaks=seq(-2, 2, length.out=101))\n",
    "\n",
    "#write.csv(log2(All_prep_heatmap+1), \"tumor_deg_mrna.csv\")\n",
    "\n",
    "\n",
    "\n",
    "####################################################\n",
    "\n",
    "#Ranking analysis\n",
    "\n",
    "#prep\n",
    "colnames(All_prep_heatmap) <- c(1:16)\n",
    "\n",
    "#Ranked genes for HumanBase\n",
    "Ranked <- log2(All_prep_heatmap+1) %>% rowwise() %>% dplyr::mutate(Ex_avg=mean(c_across(c(1,5,9,13)))) %>% \n",
    "          dplyr::mutate(EE_avg=mean(c_across(c(2,6,10,14)))) %>%\n",
    "          dplyr::mutate(FF_avg=mean(c_across(c(3,7,11,15)))) %>%\n",
    "          dplyr::mutate(UC_avg=mean(c_across(c(4,8,12,16)))) %>%\n",
    "          dplyr::mutate(Ex_rank=case_when(Ex_avg==max(c_across(c('Ex_avg','EE_avg','FF_avg','UC_avg')))~1)) %>%\n",
    "          dplyr::mutate(EE_rank=case_when(EE_avg==max(c_across(c('Ex_avg','EE_avg','FF_avg','UC_avg')))~1)) %>%\n",
    "          dplyr::mutate(FF_rank=case_when(FF_avg==max(c_across(c('Ex_avg','EE_avg','FF_avg','UC_avg')))~1)) %>%\n",
    "          dplyr::mutate(UC_rank=case_when(UC_avg==max(c_across(c('Ex_avg','EE_avg','FF_avg','UC_avg')))~1)) \n",
    "\n",
    "Ranked <- cbind(Ranked, Genes=rownames(All_prep_heatmap))\n",
    "\n",
    "arrange(Ranked, desc(Ex_avg)) %>% arrange(., desc(Ex_rank))\n",
    "\n",
    "\n",
    "####################################################################\n",
    "#Supplementary GMNN figure\n",
    "options(repr.plot.width = 12, repr.plot.height =10)\n",
    "\n",
    "GMNN <- All.heatmap['GMNN',]\n",
    "GMNN <- GMNN[,-17] \n",
    "colnames(GMNN) <- vol.metadata$Method\n",
    "labels <- colnames(GMNN)\n",
    "prep <- t(log2(GMNN+1)) %>% as.data.frame() %>% mutate(ID=rownames(.))\n",
    "prep['ID'] <- labels %>% as.factor()\n",
    "colours <- c(\"ExCy\" =\"red3\", \"ExoEasy\"=\"orange\", \"Fujifilm\"=\"grey\", \"UC\"=\"blue2\")\n",
    "#\n",
    "ggplot(prep, aes(x=GMNN, y=ID, fill=ID)) + \n",
    "    geom_boxplot(show.legend=F, alpha=0.5) +\n",
    "    geom_point(aes(color=ID, size=50), show.legend=F)+\n",
    "    scale_y_discrete(limits=rev) +\n",
    "    scale_fill_manual(values=colours) + \n",
    "    scale_color_manual(values=colours)+\n",
    "    theme_bw(base_size=25) +\n",
    "    xlab(expression(bold(GMNN ~ Log[2] ~ Transcript))) + \n",
    "    theme(\n",
    "        panel.grid.major = element_blank(),\n",
    "        panel.grid.minor=element_blank(),\n",
    "        axis.ticks.y = element_blank(),\n",
    "        axis.text.y = element_text(face=\"bold\", size=35),\n",
    "        axis.title.x=element_text(face=\"bold\", size=35)\n",
    "        ) + \n",
    "    ylab(element_blank()) \n",
    "    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "UFRC R-4.1",
   "language": "R",
   "name": "ir-ufrc-4.1"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.1.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
