{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "9b5e2d6f-0ca1-4e7f-876b-dd772d23ccf7",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#This block to load all libraries needed\n",
    "library(\"DEXSeq\")\n",
    "library(\"tidyr\")\n",
    "library(\"qvalue\")\n",
    "library(\"edgeR\")\n",
    "library(\"dplyr\")\n",
    "library(\"reshape2\")\n",
    "library(\"EnhancedVolcano\")\n",
    "library(\"sva\")\n",
    "library(\"Rtsne\")\n",
    "library(\"ggplot2\")\n",
    "library(\"pheatmap\")\n",
    "library(\"org.Hs.eg.db\")\n",
    "library(\"readr\")\n",
    "library('scales')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "36554c8f-aa1b-4ec1-b583-d00dabd168f0",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "converting counts to integer mode\n",
      "\n",
      "Warning message in DESeqDataSet(rse, design, ignoreRank = TRUE):\n",
      "“some variables in design formula are characters, converting to factors”\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "Warning message:\n",
      "“attributes are not identical across measure variables; they will be dropped”\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'Gene'. You can override using the\n",
      "`.groups` argument.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m9989\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m1\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (1): mRNA\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene_id'. You can override using the\n",
      "`.groups` argument.\n"
     ]
    }
   ],
   "source": [
    "##GFF\n",
    "flattenedFile = list.files('.', pattern=\"gff$\", full.names=TRUE)\n",
    "\n",
    "#METADATA\n",
    "metadata = read.csv('ExCy_metadata_dexseq.csv')\n",
    "rownames(metadata) <- metadata$SampleID\n",
    "metadata <- metadata[,-1]\n",
    "\n",
    "#Entire dataset generation here to do downstream analyses\n",
    "All.countfiles=list.files('./ALL/', pattern=\"txt$\", full.names=TRUE)\n",
    "\n",
    "##Generate count file\n",
    "All.dxd = DEXSeqDataSetFromHTSeq(countfiles = All.countfiles, sampleData = metadata[c(1:28),], design = ~ sample + exon, flattenedfile = flattenedFile )\n",
    "All.count <- as.data.frame(featureCounts(All.dxd))\n",
    "\n",
    "#Do the light filtering and aggregation\n",
    "All.count$ID <- rownames(All.count)\n",
    "All.count <- All.count %>% mutate(Gene = substr(ID, 1,15))\n",
    "All.cf <- All.count %>% mutate(row_sum = rowSums(.[,-29:-30])) %>% dplyr::filter(row_sum > 10) %>% select(-ID)\n",
    "All.cfs <-  melt(All.cf, id.vars = \"Gene\") %>% group_by(Gene, variable) %>% summarise(sum_value=sum(value)) %>% dcast(Gene ~ variable, value.var=\"sum_value\")\n",
    "rownames(All.cfs) <- All.cfs$Gene\n",
    "\n",
    "\n",
    "#prepare for annotation\n",
    "exomega <- as.data.frame(read.csv('ExoMEGA.txt'))\n",
    "mRNA <- read_tsv('VESICLEPEDIA_PROTEIN_MRNAS_5.1.txt')\n",
    "\n",
    "##map gene names to HUGO\n",
    "annot.df <- mapIds(org.Hs.eg.db, keys = row.names(All.cfs), column = \"SYMBOL\", keytype = \"ENSEMBL\") %>% as.data.frame() %>% rename(SYMBOL=1)\n",
    "annot.df$SYMBOL <- toupper(annot.df$SYMBOL)\n",
    "All.cfs$gene_id <- annot.df$SYMBOL \n",
    "\n",
    "#filter out non-mapped genes and combine duplicate\n",
    "All.gcfs <- All.cfs %>% dplyr::filter(!grepl(\"NA\", gene_id, fixed= TRUE)) %>% select(-Gene)\n",
    "All.gpcfs <-  melt(All.gcfs, id.vars = \"gene_id\") %>% group_by(gene_id, variable) %>% summarise(sum_value=sum(value)) %>% dcast(gene_id ~ variable, value.var=\"sum_value\")\n",
    "All.gpcfs <- All.gpcfs[complete.cases(All.gpcfs), ]\n",
    "\n",
    "\n",
    "All.exomega <- as.data.frame(All.gpcfs[All.gpcfs$gene_id %in% exomega[,1] ,])\n",
    "rownames(All.exomega) <- All.exomega$gene_id\n",
    "mRNA_All <- All.exomega[rownames(All.exomega) %in% mRNA$mRNA ,]\n",
    "All.exomega <- mRNA_All[,2:29]\n",
    "\n",
    "#write.csv(All.gpcfs, 'TOTAL_RNA.csv')\n",
    "\n",
    "####################################################\n",
    "\n",
    "#ALL\n",
    "\n",
    "#edgeR\n",
    "vol.metadata <- metadata\n",
    "status <- relevel(factor(vol.metadata$Status), ref=\"Healthy\")\n",
    "design <- model.matrix(~status)\n",
    "d <- DGEList(counts = All.exomega)\n",
    "d <- calcNormFactors(d)\n",
    "d <- estimateDisp(d, design, robust=TRUE)\n",
    "d_fit <- glmFit(d, design)\n",
    "qlf <- glmLRT(d_fit)\n",
    "qlf_table <- topTags(qlf, n=Inf)$table\n",
    "write.csv(qlf_table, 'All_Tumor_vs_healthy.csv')\n",
    "#head(qlf_table)\n",
    "\n",
    "\n",
    "####################################################\n",
    "\n",
    "#ExCy\n",
    "vol.metadata <- metadata[c(1,5,9,13,17,21,25),]\n",
    "status <- relevel(factor(vol.metadata$Status), ref=\"Healthy\")\n",
    "design <- model.matrix(~status)\n",
    "d <- DGEList(counts = All.exomega[,c(1,5,9,13,17,21,25)])\n",
    "d <- calcNormFactors(d)\n",
    "d <- estimateDisp(d, design, robust=TRUE)\n",
    "d_fit <- glmFit(d, design)\n",
    "qlf <- glmLRT(d_fit)\n",
    "qlf_table <- topTags(qlf, n=Inf)$table\n",
    "write.csv(qlf_table, 'ExCy_Tumor_vs_healthy.csv')\n",
    "\n",
    "\n",
    "\n",
    "####################################################\n",
    "\n",
    "#ExoEasy\n",
    "vol.metadata <- metadata[c(2,6,10,14,18,22,26),]\n",
    "status <- relevel(factor(vol.metadata$Status), ref=\"Healthy\")\n",
    "design <- model.matrix(~status)\n",
    "d <- DGEList(counts = All.exomega[,c(2,6,10,14,18,22,26)])\n",
    "d <- calcNormFactors(d)\n",
    "d <- estimateDisp(d, design, robust=TRUE)\n",
    "d_fit <- glmFit(d, design)\n",
    "qlf <- glmLRT(d_fit)\n",
    "qlf_table <- topTags(qlf, n=Inf)$table\n",
    "write.csv(qlf_table, 'EE_Tumor_vs_healthy.csv')\n",
    "\n",
    "####################################################\n",
    "\n",
    "#FF\n",
    "vol.metadata <- metadata[c(3,7,11,15,19,23,27),]\n",
    "status <- relevel(factor(vol.metadata$Status), ref=\"Healthy\")\n",
    "design <- model.matrix(~status)\n",
    "d <- DGEList(counts = All.exomega[,c(3,7,11,15,19,23,27)])\n",
    "d <- calcNormFactors(d)\n",
    "d <- estimateDisp(d, design, robust=TRUE)\n",
    "d_fit <- glmFit(d, design)\n",
    "qlf <- glmLRT(d_fit)\n",
    "qlf_table <- topTags(qlf, n=Inf)$table\n",
    "write.csv(qlf_table, 'FF_Tumor_vs_healthy.csv')\n",
    "\n",
    "####################################################\n",
    "\n",
    "#UC\n",
    "vol.metadata <- metadata[c(4,8,12,16,20,24,28),]\n",
    "status <- relevel(factor(vol.metadata$Status), ref=\"Healthy\")\n",
    "design <- model.matrix(~status)\n",
    "d <- DGEList(counts = All.exomega[,c(4,8,12,16,20,24,28)])\n",
    "d <- calcNormFactors(d)\n",
    "d <- estimateDisp(d, design, robust=TRUE)\n",
    "d_fit <- glmFit(d, design)\n",
    "qlf <- glmLRT(d_fit)\n",
    "qlf_table <- topTags(qlf, n=Inf)$table\n",
    "write.csv(qlf_table, 'UC_Tumor_vs_healthy.csv')\n",
    "\n",
    "####################################################\n",
    "\n",
    "#Summary of results down below. We expand the fdr to 0.10 since small dataset. \n",
    "\n",
    "#ALL - MOG, FBL\n",
    "\n",
    "#ExoEasy - ZFPL1, NDUFB11, \n",
    "\n",
    "#FF -  NT5C, RASF7, BAG6, COX6A2, ACR, IGKV1D-33\n",
    "\n",
    "#ExCy - ATP6V0B\n",
    "\n",
    "#UC - PRPS2, ARHGDIA, ERAS, LILRB3, S100A8, TMC6, NIT1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "34cf36ba-dc38-45bb-8cb3-5edbc9326536",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "converting counts to integer mode\n",
      "\n",
      "Warning message in DESeqDataSet(rse, design, ignoreRank = TRUE):\n",
      "“some variables in design formula are characters, converting to factors”\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "  Note: levels of factors in the design contain characters other than\n",
      "  letters, numbers, '_' and '.'. It is recommended (but not required) to use\n",
      "  only letters, numbers, and delimiters '_' or '.', as these are safe characters\n",
      "  for column names in R. [This is a message, not a warning or an error]\n",
      "\n",
      "Warning message:\n",
      "“attributes are not identical across measure variables; they will be dropped”\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'Gene'. You can override using the\n",
      "`.groups` argument.\n",
      "\u001b[1mRows: \u001b[22m\u001b[34m9989\u001b[39m \u001b[1mColumns: \u001b[22m\u001b[34m1\u001b[39m\n",
      "\u001b[36m──\u001b[39m \u001b[1mColumn specification\u001b[22m \u001b[36m────────────────────────────────────────────────────────\u001b[39m\n",
      "\u001b[1mDelimiter:\u001b[22m \"\\t\"\n",
      "\u001b[31mchr\u001b[39m (1): mRNA\n",
      "\n",
      "\u001b[36mℹ\u001b[39m Use `spec()` to retrieve the full column specification for this data.\n",
      "\u001b[36mℹ\u001b[39m Specify the column types or set `show_col_types = FALSE` to quiet this message.\n",
      "'select()' returned 1:many mapping between keys and columns\n",
      "\n",
      "\u001b[1m\u001b[22m`summarise()` has grouped output by 'gene_id'. You can override using the\n",
      "`.groups` argument.\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 6 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th></th><th scope=col>logFC</th><th scope=col>logCPM</th><th scope=col>LR</th><th scope=col>PValue</th><th scope=col>FDR</th></tr>\n",
       "\t<tr><th></th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><th scope=row>ATP6V0B</th><td> 2.024615</td><td>6.569898</td><td>19.15903</td><td>1.202674e-05</td><td>0.09608159</td></tr>\n",
       "\t<tr><th scope=row>ICOSLG</th><td> 4.480779</td><td>3.101811</td><td>16.69810</td><td>4.382494e-05</td><td>0.17505872</td></tr>\n",
       "\t<tr><th scope=row>IRF7</th><td> 8.026398</td><td>2.894858</td><td>14.90488</td><td>1.130700e-04</td><td>0.22986096</td></tr>\n",
       "\t<tr><th scope=row>HRK</th><td> 1.600193</td><td>6.807256</td><td>14.24983</td><td>1.600749e-04</td><td>0.22986096</td></tr>\n",
       "\t<tr><th scope=row>PSMB4</th><td>-1.947523</td><td>6.047434</td><td>13.87947</td><td>1.949164e-04</td><td>0.22986096</td></tr>\n",
       "\t<tr><th scope=row>MPHOSPH10</th><td>-1.427288</td><td>6.650630</td><td>13.62371</td><td>2.233463e-04</td><td>0.22986096</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 6 × 5\n",
       "\\begin{tabular}{r|lllll}\n",
       "  & logFC & logCPM & LR & PValue & FDR\\\\\n",
       "  & <dbl> & <dbl> & <dbl> & <dbl> & <dbl>\\\\\n",
       "\\hline\n",
       "\tATP6V0B &  2.024615 & 6.569898 & 19.15903 & 1.202674e-05 & 0.09608159\\\\\n",
       "\tICOSLG &  4.480779 & 3.101811 & 16.69810 & 4.382494e-05 & 0.17505872\\\\\n",
       "\tIRF7 &  8.026398 & 2.894858 & 14.90488 & 1.130700e-04 & 0.22986096\\\\\n",
       "\tHRK &  1.600193 & 6.807256 & 14.24983 & 1.600749e-04 & 0.22986096\\\\\n",
       "\tPSMB4 & -1.947523 & 6.047434 & 13.87947 & 1.949164e-04 & 0.22986096\\\\\n",
       "\tMPHOSPH10 & -1.427288 & 6.650630 & 13.62371 & 2.233463e-04 & 0.22986096\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 6 × 5\n",
       "\n",
       "| <!--/--> | logFC &lt;dbl&gt; | logCPM &lt;dbl&gt; | LR &lt;dbl&gt; | PValue &lt;dbl&gt; | FDR &lt;dbl&gt; |\n",
       "|---|---|---|---|---|---|\n",
       "| ATP6V0B |  2.024615 | 6.569898 | 19.15903 | 1.202674e-05 | 0.09608159 |\n",
       "| ICOSLG |  4.480779 | 3.101811 | 16.69810 | 4.382494e-05 | 0.17505872 |\n",
       "| IRF7 |  8.026398 | 2.894858 | 14.90488 | 1.130700e-04 | 0.22986096 |\n",
       "| HRK |  1.600193 | 6.807256 | 14.24983 | 1.600749e-04 | 0.22986096 |\n",
       "| PSMB4 | -1.947523 | 6.047434 | 13.87947 | 1.949164e-04 | 0.22986096 |\n",
       "| MPHOSPH10 | -1.427288 | 6.650630 | 13.62371 | 2.233463e-04 | 0.22986096 |\n",
       "\n"
      ],
      "text/plain": [
       "          logFC     logCPM   LR       PValue       FDR       \n",
       "ATP6V0B    2.024615 6.569898 19.15903 1.202674e-05 0.09608159\n",
       "ICOSLG     4.480779 3.101811 16.69810 4.382494e-05 0.17505872\n",
       "IRF7       8.026398 2.894858 14.90488 1.130700e-04 0.22986096\n",
       "HRK        1.600193 6.807256 14.24983 1.600749e-04 0.22986096\n",
       "PSMB4     -1.947523 6.047434 13.87947 1.949164e-04 0.22986096\n",
       "MPHOSPH10 -1.427288 6.650630 13.62371 2.233463e-04 0.22986096"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "##GFF\n",
    "flattenedFile = list.files('.', pattern=\"gff$\", full.names=TRUE)\n",
    "\n",
    "#METADATA\n",
    "metadata = read.csv('ExCy_metadata_dexseq.csv')\n",
    "rownames(metadata) <- metadata$SampleID\n",
    "metadata <- metadata[,-1]\n",
    "\n",
    "#Entire dataset generation here to do downstream analyses\n",
    "All.countfiles=list.files('./ALL/', pattern=\"txt$\", full.names=TRUE)\n",
    "\n",
    "##Generate count file\n",
    "All.dxd = DEXSeqDataSetFromHTSeq(countfiles = All.countfiles, sampleData = metadata[c(1:28),], design = ~ sample + exon, flattenedfile = flattenedFile )\n",
    "All.count <- as.data.frame(featureCounts(All.dxd))\n",
    "\n",
    "#Do the light filtering and aggregation\n",
    "All.count$ID <- rownames(All.count)\n",
    "All.count <- All.count %>% mutate(Gene = substr(ID, 1,15))\n",
    "All.cf <- All.count %>% mutate(row_sum = rowSums(.[,-29:-30])) %>% dplyr::filter(row_sum > 10) %>% select(-ID)\n",
    "All.cfs <-  melt(All.cf, id.vars = \"Gene\") %>% group_by(Gene, variable) %>% summarise(sum_value=sum(value)) %>% dcast(Gene ~ variable, value.var=\"sum_value\")\n",
    "rownames(All.cfs) <- All.cfs$Gene\n",
    "\n",
    "\n",
    "#prepare for annotation\n",
    "exomega <- as.data.frame(read.csv('ExoMEGA.txt'))\n",
    "mRNA <- read_tsv('VESICLEPEDIA_PROTEIN_MRNAS_5.1.txt')\n",
    "\n",
    "##map gene names to HUGO\n",
    "annot.df <- mapIds(org.Hs.eg.db, keys = row.names(All.cfs), column = \"SYMBOL\", keytype = \"ENSEMBL\") %>% as.data.frame() %>% rename(SYMBOL=1)\n",
    "annot.df$SYMBOL <- toupper(annot.df$SYMBOL)\n",
    "All.cfs$gene_id <- annot.df$SYMBOL \n",
    "\n",
    "#filter out non-mapped genes and combine duplicate\n",
    "All.gcfs <- All.cfs %>% dplyr::filter(!grepl(\"NA\", gene_id, fixed= TRUE)) %>% select(-Gene)\n",
    "All.gpcfs <-  melt(All.gcfs, id.vars = \"gene_id\") %>% group_by(gene_id, variable) %>% summarise(sum_value=sum(value)) %>% dcast(gene_id ~ variable, value.var=\"sum_value\")\n",
    "All.gpcfs <- All.gpcfs[complete.cases(All.gpcfs), ]\n",
    "\n",
    "\n",
    "All.exomega <- as.data.frame(All.gpcfs[All.gpcfs$gene_id %in% exomega[,1] ,])\n",
    "rownames(All.exomega) <- All.exomega$gene_id\n",
    "mRNA_All <- All.exomega[rownames(All.exomega) %in% mRNA$mRNA ,]\n",
    "All.exomega <- mRNA_All[,2:29]\n",
    "#ExCy\n",
    "vol.metadata <- metadata[c(1,5,9,13,17,21,25),]\n",
    "status <- relevel(factor(vol.metadata$Status), ref=\"Healthy\")\n",
    "design <- model.matrix(~status)\n",
    "d <- DGEList(counts = All.exomega[,c(1,5,9,13,17,21,25)])\n",
    "d <- calcNormFactors(d)\n",
    "d <- estimateDisp(d, design, robust=TRUE)\n",
    "d_fit <- glmFit(d, design)\n",
    "qlf <- glmLRT(d_fit)\n",
    "qlf_table <- topTags(qlf, n=Inf)$table\n",
    "head(qlf_table)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "87b44b2d-c922-4e6c-accc-f81741544e9c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "UFRC R-4.2",
   "language": "R",
   "name": "ir-ufrc-4.2"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
